connectome_interpreter/utils.py

```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+
+    # check if length of v1 and v2 are the same
+    if len(v1) != len(v2):
+        raise ValueError("The length of v1 and v2 should be the same.")
+
+    if device is None:
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    sc1 = np.linspace(0, 1, grid_size)
+    sc2 = np.linspace(0, 1, grid_size)
+    sc1_grid, sc2_grid = np.meshgrid(sc1, sc2)
+
+    inputs = np.zeros((grid_size**2, len(v1), num_layers), dtype=np.float32)
+    for i, (s1, s2) in enumerate(zip(sc1_grid.ravel(), sc2_grid.ravel())):
+        inputs[i, :, timepoints] = v1 * s1 + v2 * s2
+
+    # Convert to torch tensor
+    inputs_tensor = torch.from_numpy(inputs).to(device)
+
+    # Create grid coordinates for reference
+    grid_coords = list(zip(sc1_grid.ravel(), sc2_grid.ravel()))
+
+    return inputs_tensor, grid_coords
+
+
+def plot_output_grid(grid_outputs: torch.Tensor | npt.NDArray, grid_coords: List[Tuple[float, float]], selected_index: int | List[int], figsize: tuple = (10, 8),
+                     cmap: str = 'viridis', xlab: str = 'v1 Activation', ylab: str = 'v2 Activation', title: str = None):
+    """
+    Plot the output grid for selected indices.
+
+    Args:
+        grid_outputs (torch.Tensor|np.ndarray): The output grid tensor of shape (grid_size**2, num_neurons, num_layers).
+        grid_coords (list): A list of grid coordinates. Best from function `make_grid_inputs()`.
+        selected_index (int|list): The index or indices to plot.
+        figsize (tuple, optional): The size of the figure. Defaults to (10, 8).
+    """
+    selected_index = to_nparray(selected_index)
+
+    # if grid_outputs is tensor, convert to numpy
+    if isinstance(grid_outputs, torch.Tensor):
+        grid_outputs = grid_outputs.cpu().numpy()
+
+    if title is None:
+        title = f'Output Grid for {selected_index}'
+
+    def plot_heatmap(index, xlab, ylab, title):
+        # # Create the plot
+        plt.figure(figsize=figsize)
+        im = plt.imshow(heatmaps[index-1], cmap=cmap, origin='lower',
+                        extent=[0, 1, 0, 1])
+        plt.colorbar(im, label='Output')
+        plt.xlabel(xlab)
+        plt.ylabel(ylab)
+        plt.title(title)
+        plt.show()
+
+    heatmaps = []
+
+    for i in range(grid_outputs.shape[2]):
+        # Extract the selected index values
+        values = grid_outputs[:, selected_index,
+                              i].mean(axis=1)
+
+        # Reshape values to match the grid
+        grid_size = int(np.sqrt(len(grid_coords)))
+        values_grid = values.reshape((grid_size, grid_size))
+        heatmaps.append(values_grid)
+
+    slider = widgets.IntSlider(value=1, min=1, max=len(heatmaps), step=1,
+                               description='Timepoint', continuous_update=True)
+
+    # Link the slider to the plotting function
+    display(widgets.interactive(plot_heatmap, index=slider,
+            xlab=xlab, ylab=ylab, title=title))
+
+
+@dataclass
```

Comment:
Does this stuff want to be in a separate module? 

I try to avoid having a `utils.py` module that grows too large; business logic like this probably deserves to be in its own module. I try to reserve `utils` for silly low-level things like string formatting or routine calculations.

Reply:
Good point, moved.